### ISSUE TITLE
invalid operation - Zip4j is in busy state

### ISSUE THREAD

**Issue by DeflatedPickle** [srikanth-lingala/zip4j#322]
I've been using Zip4J to unzip contents from a zipped file and it works flawlessly however upon running my `ZipFile` in a thread to make use of the `ProgressMonitor`, I'm getting the error in the title as a result of [this line](https://github.com/srikanth-lingala/zip4j/blob/cd37bb56ef3127e18b270d82f8f23fd5a56386fb/src/main/java/net/lingala/zip4j/ZipFile.java#L680) whenever I try to call `extractFile`

My (Kotlin) code is as follows;
```kotlin
val zipFile = ZipFile(file).apply { 
    isRunInThread = true
}

for (i in zipFile.fileHeaders.filter {
    it.fileName.startsWith("assets/minecraft")
}) {
    if (!i.isDirectory) {
        zipFile.extractFile(
            i.fileName,
            path
        )
    }
}
```
If I don't run it in a thread, it works as expected

**Comment by srikanth-lingala** [comment]
Zip4j thread mode uses a SingleThreadPoolExecutor. It does its "actions" (extracting, adding, etc) in a single thread when you run it in a thread mode (runInThread = true). When there is an action going on, you need to wait until that action is complete before you perform a next action. In this case, each extraction of a file is an action. You have to check for the status of the execution and perform a next action only when zip4j is no more in busy state. Something like this should work (I included only the relevant parts of your code, and my sample is in Java):

```
if (!i.isDirectory) {
    zipFile.extractFile(
        i.fileName,
        path
    )

  while (!progressMonitor.getState().equals(ProgressMonitor.State.READY)) {
    Thread.sleep(100);
  }

    if (progressMonitor.getResult().equals(ProgressMonitor.Result.ERROR)) {
      System.out.println("Error occurred extracting file. Error message: " + progressMonitor.getException().getMessage());
    } else if (progressMonitor.getResult().equals(ProgressMonitor.Result.CANCELLED)) {
      System.out.println("Task cancelled");
    }
}
```

Basically, you have to wait until the current action is complete and the state returns to `READY` before you can perform the next file extraction. You probably also want to check the status of each extraction before you continue, so you don't miss any errors that occurred. Please have a look at the documentation [here](https://github.com/srikanth-lingala/zip4j#working-with-progress-monitor).

**Comment by DeflatedPickle** [comment]
So, that code snippet simply doesn't work with a GUI program as it pauses the main thread. Using Kotlin's coroutines sometimes works, though Zip4J can report being in a busy state depending on where the coroutine is placed, but it can also happen when running it in a different thread. It feels like either a coroutine or a thread should be used so the UI doesn't freeze

This isn't the worst issue however as the following code snippet;
```kotlin
val zipFile = ZipFile(file).apply {
    isRunInThread = true
}

val progressMonitor = zipFile.progressMonitor

for (
    i in zipFile.fileHeaders.filter {
        it.fileName.startsWith("assets/minecraft")
    }
) {
    if (!i.isDirectory) {
        zipFile.extractFile(
            i.fileName,
            path
        )

        while (progressMonitor.state != ProgressMonitor.State.READY) {
            println(progressMonitor.percentDone)
        }

        when (progressMonitor.result) {
            ProgressMonitor.Result.ERROR ->
                logger.error("Pack extraction ran into the error; ${progressMonitor.exception.message}")
            ProgressMonitor.Result.CANCELLED ->
                logger.warn("User cancelled pack extraction")
            else -> {}
        }
    }
}
```
Gives the following output;
```
0
0
0
0
0
0
0
0
0
0
0
100
0
0
0
0
0
0
0
0
0
0
0
0
0
0
```
Which is hardly useful when tracking progress. It isn't even always that. Sometimes it goes back and forth between `0` and `100` during extraction, but it never gives another number

**Comment by srikanth-lingala** [comment]
> So, that code snippet simply doesn't work with a GUI program as it pauses the main thread.

It depends on your implementation I guess. The reason for thread mode in zip4j is exactly that - to not block the main thread. Zip4j runs in a different thread making your main thread available for your processing. A lot of Android apps and GUI applications use zip4j with thread mode. I can't go into the implementation details, but all I can say is that your main thread is not blocked when running in thread mode and is available for your application.

As for the progress status going from 0 to 100 to 0: I believe this is a behaviour of too many parallel writes to sysout within a short span of time. I would recommend adding a small `Thread.sleep()` of at least 50 to 100 millis in your while loop and I believe you will not see this behaviour.

And the reason why it goes directly from 0 to 100 without any intermediate numbers is that this depends on a lot on the size of the file being extracted and the processing power of the machine where zip4j is running. I think that the files being extracted are too small in size to see any noticeable percentage change. Use a bigger file size and you should see numbers gradually increasing.

**Comment by DeflatedPickle** [comment]
> The reason for thread mode in zip4j is exactly that - to not block the main thread

Maybe this is my fault for not using a `SwingWorker`, but even when in the threaded mode, Zip4J freezes my GUI

> I would recommend adding a small Thread.sleep() of at least 50 to 100 millis in your while loop

I did try adding a small sleep as all the examples have one but I saw similar output, just with pauses

> And the reason why it goes directly from 0 to 100 without any intermediate numbers is that this depends on a lot on the size of the file being extracted and the processing power of the machine where zip4j is running

This is supposed to be a percentage though? Wouldn't it make sense that the total file count was requested first, then when each file was extracted, add to the current value. Or if it isn't possible to check when an individual file is extracted, routinely check the destination directory and compare the total files in that to the total in the ZIP?

> Use a bigger file size and you should see numbers gradually increasing

This is a very silly suggestion

**Comment by srikanth-lingala** [comment]
> This is a very silly suggestion

First of all, I would appreciate if you show a bit of respect around here. I put in a lot of effort to maintain develop and maintain this library for free of use by the community. If you like it, use it. If not, try to find any alternatives that works for you.

With that said: When I said use bigger files, I meant as a demonstration to see the gradual progression of percentage, but not in your production code of course. Using a bigger file size in your test will show you that the percentage does not always go directly from 0 to 100, but gradually increases. If your file size being extracted is too small for the processing power of your machine, by the time you query for the percentage done again, the file is already extracted, and therefore the percentage is 100. How do you expect zip4j to behave? Add a pause just so that user sees percentage gradually increase? 

> but even when in the threaded mode, Zip4J freezes my GUI

 I repeat: this is your implementation details. Like I repeated several times in my previous post: Zip4j when running in thread mode does not block your main thread. If your GUI is still blocked, it is your implementation details. As I said, zip4j is already used by many Android apps and GUI applications in thread mode and I never heard any issues from them.

> I did try adding a small sleep as all the examples have one but I saw similar output, just with pauses

Post your standalone code along with a zip file here, and I can try to help

> This is supposed to be a percentage though? Wouldn't it make sense that the total file count was requested first, then when each file was extracted, add to the current value. Or if it isn't possible to check when an individual file is extracted, routinely check the destination directory and compare the total files in that to the total in the ZIP?

Zip4j does something like this. It initially calculates the total amount of work to be done, and keeps adding work that is done to it, so when you query for percentage done, it is a simple calculation based on those two values. Listen, your question is like asking why don't I see progress bar in a Windows unzip GUi go gradually to 100% when extracting a 10kb zip file. In this case you will see the percentage jump to 100% directly (unless you use a really slow machine), because by the time the GUI updates the percentage, the file extraction is already done. Something similar is happening here as well. As I said earlier, in order to prove that your percentage gradually increases, you can use a bigger file size or a slower machine (again just for demo purposes and in test code and I don't mean not in production).

Good luck.
